server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -337,6 +338,29 @@ Runnable getGlobalCheckpointSyncer() {
 
     private final List<ReferenceManager.RefreshListener> internalRefreshListener = new ArrayList<>();
 
+    /**
+     * An OpenSearch index shard config supplier
+     *
+     * @opensearch.internal
+     */
+    static public class IndexShardConfigSupplier {
```

Comment:
The structure seems to be force-fitting this very specific use case. We should model it more generically